pkg/controller/bastion/options.go

```diff
@@ -0,0 +1,81 @@
+// Copyright (c) 2021 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package bastion
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/gardener/gardener-extension-provider-gcp/pkg/gcp"
+	"github.com/gardener/gardener/extensions/pkg/controller"
+	gardencorev1beta1 "github.com/gardener/gardener/pkg/apis/core/v1beta1"
+	extensionsv1alpha1 "github.com/gardener/gardener/pkg/apis/extensions/v1alpha1"
+	corev1 "k8s.io/api/core/v1"
+)
+
+// Options contains provider-related information required for setting up
+// a bastion instance. This struct combines precomputed values like the
+// bastion instance name with the IDs of pre-existing cloud provider
+// resources, like the VPC ID, subnet ID etc.
+type Options struct {
+	Shoot               *gardencorev1beta1.Shoot
+	BastionInstanceName string
+	FirewallName        string
+	PrivateIP           string
+	PublicIP            string
+	DiskName            string
+	UserData            string
+	Zone                string
+	Region              string
+	SubnetWork          string
+	ProjectID           string
+}
+
+// DetermineOptions determines the required information that are required to reconcile a Bastion on GCP. This
+// function does not create any IaaS resources.
+func DetermineOptions(ctx context.Context, bastion *extensionsv1alpha1.Bastion, cluster *controller.Cluster) (*Options, error) {
+	name := cluster.ObjectMeta.Name
+	bastionInstanceName := fmt.Sprintf("%s-%s-bastion", name, bastion.Name)
+	firewallName := fmt.Sprintf("%s-allow-ssh-access", bastionInstanceName)
+	diskName := fmt.Sprintf("%s-%s-disk", name, bastion.Name)
+	publicIP := bastion.Spec.Ingress[0].IPBlock.CIDR
+	userData := string(bastion.Spec.UserData)
+	subnetwork := cluster.Shoot.Name + "-nodes"
+	zone := strings.Join(cluster.Shoot.Spec.Provider.Workers[0].Zones, " ")
```

Comment:
1) > What if the worker pools change completely (especially if the zone of the first worker pool is different) after a bastion was created? will you still be able to find the created resources in order to clean them up afterwards?

Thanks, point that. it could be. an SSH session sticks in the problematic node for troubleshooting, but timeout to join a cluster then drops the problematic node. I will think about how to handle it

Reply:
fixed in d5a96db004a9bfcb41aaaa41cc695a0fd062817f

Reply:
The first issue is still not resolved

Reply:
oh, sorry wrong PR I pasted, here is the correct one 45b6ab860e98a216739943e0e4e6eaf3f63fb05e. alwasy use first available zone from `cluster.CloudProfile`.

Reply:
```go
func getZone(cluster *extensions.Cluster, region string) string {
	for _, j := range cluster.CloudProfile.Spec.Regions {
		if j.Name == region {
			return j.Zones[0].Name
		}
	}
	return ""
}
```
You now take the first zone of the region from the Cloudprofile, but this list is also not guaranteed to be stable, so you have the same problem as before. You need to store the information with which zone you have created the resources, otherwise you potentially leak resources.

I talked with @rfranzke, he suggested to store the value under `.status.providerStatus` https://gardener.cloud/docs/references/extensions/#extensions.gardener.cloud/v1alpha1.DefaultStatus
This way, on subsequent reconciles or on deletion you will get the "initial" zone and you can be sure to find the resources that you have created.

Reply:
fixed 31b6040  and 3064ac3